Error on Display Owner

Caused by: org.thymeleaf.exceptions.TemplateProcessingException: Exception evaluating SpringEL expression: "pet.type" (template: "owners/ownerDetails" - line 52, col 21)
	at org.thymeleaf.spring5.expression.SPELVariableExpressionEvaluator.evaluate(SPELVariableExpressionEvaluator.java:292) ~[thymeleaf-spring5-3.0.15.RELEASE.jar:3.0.15.RELEASE]
	at org.thymeleaf.standard.expression.VariableExpression.executeVariableExpression(VariableExpression.java:166) ~[thymeleaf-3.0.15.RELEASE.jar:3.0.15.RELEASE]
	at org.thymeleaf.standard.expression.SimpleExpression.executeSimple(SimpleExpression.java:66) ~[thymeleaf-3.0.15.RELEASE.jar:3.0.15.RELEASE]
	at org.thymeleaf.standard.expression.Expression.execute(Expression.java:109) ~[thymeleaf-3.0.15.RELEASE.jar:3.0.15.RELEASE]
	at org.thymeleaf.standard.expression.Expression.execute(Expression.java:138) ~[thymeleaf-3.0.15.RELEASE.jar:3.0.15.RELEASE]

--- a/pet-clinic-data/src/main/java/com/osyrs/sfgpetclinic/model/PetType.java
+++ b/pet-clinic-data/src/main/java/com/osyrs/sfgpetclinic/model/PetType.java
@@ -16,4 +16,9 @@ import javax.persistence.Table;
 public class PetType extends BaseEntity{
     @Column(name = "name")
     private String name;
+
+    @Override
+    public String toString() {
+        return name;
+    }
 }

--- a/pet-clinic-web/src/main/java/com/osyrs/sfgpetclinic/controllers/OwnerController.java
+++ b/pet-clinic-web/src/main/java/com/osyrs/sfgpetclinic/controllers/OwnerController.java
@@ -3,7 +3,10 @@ package com.osyrs.sfgpetclinic.controllers;
 import com.osyrs.sfgpetclinic.services.OwnerService;
 import org.springframework.stereotype.Controller;
 import org.springframework.ui.Model;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
 import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.servlet.ModelAndView;
 
 @Controller
 @RequestMapping("/owners")
@@ -23,5 +26,12 @@ public class OwnerController {
     @RequestMapping("/find")
     public String findOwners(){
         return "notimplemented";
+    }
+
+    @GetMapping("/{ownerId}")
+    public ModelAndView showOwner(@PathVariable("ownerId") Long ownerId){
+        var mav = new ModelAndView("owners/ownerDetails");
+        mav.addObject(ownerService.findById(ownerId));
+        return mav;
     }
 }

--- a/pet-clinic-web/src/main/resources/templates/owners/createOrUpdateOwnerForm.html
+++ b/pet-clinic-web/src/main/resources/templates/owners/createOrUpdateOwnerForm.html
@@ -1,0 +1,30 @@
+<html xmlns:th="https://www.thymeleaf.org"
+      th:replace="~{fragments/layout :: layout (~{::body},'owners')}">
+
+<body>
+
+<h2>Owner</h2>
+<form th:object="${owner}" class="form-horizontal" id="add-owner-form" method="post">
+    <div class="form-group has-feedback">
+        <input
+                th:replace="~{fragments/inputField :: input ('First Name', 'firstName', 'text')}" />
+        <input
+                th:replace="~{fragments/inputField :: input ('Last Name', 'lastName', 'text')}" />
+        <input
+                th:replace="~{fragments/inputField :: input ('Address', 'address', 'text')}" />
+        <input
+                th:replace="~{fragments/inputField :: input ('City', 'city', 'text')}" />
+        <input
+                th:replace="~{fragments/inputField :: input ('Telephone', 'telephone', 'text')}" />
+    </div>
+    <div class="form-group">
+        <div class="col-sm-offset-2 col-sm-10">
+            <button
+                    th:with="text=${owner['new']} ? 'Add Owner' : 'Update Owner'"
+                    class="btn btn-primary" type="submit" th:text="${text}">Add
+                Owner</button>
+        </div>
+    </div>
+</form>
+</body>
+</html>

--- a/pet-clinic-web/src/main/resources/templates/owners/findOwners.html
+++ b/pet-clinic-web/src/main/resources/templates/owners/findOwners.html
@@ -1,0 +1,34 @@
+<html xmlns:th="https://www.thymeleaf.org"
+      th:replace="~{fragments/layout :: layout (~{::body},'owners')}">
+
+<body>
+
+<h2>Find Owners</h2>
+
+<form th:object="${owner}" th:action="@{/owners}" method="get"
+      class="form-horizontal" id="search-owner-form">
+    <div class="form-group">
+        <div class="control-group" id="lastNameGroup">
+            <label class="col-sm-2 control-label">Last name </label>
+            <div class="col-sm-10">
+                <input class="form-control" th:field="*{lastName}" size="30"
+                       maxlength="80" /> <span class="help-inline"><div
+                    th:if="${#fields.hasAnyErrors()}">
+              <p th:each="err : ${#fields.allErrors()}" th:text="${err}">Error</p>
+            </div></span>
+            </div>
+        </div>
+    </div>
+    <div class="form-group">
+        <div class="col-sm-offset-2 col-sm-10">
+            <button type="submit" class="btn btn-primary">Find
+                Owner</button>
+        </div>
+    </div>
+
+    <a class="btn btn-primary" th:href="@{/owners/new}">Add Owner</a>
+
+</form>
+
+</body>
+</html>

--- a/pet-clinic-web/src/main/resources/templates/owners/ownerDetails.html
+++ b/pet-clinic-web/src/main/resources/templates/owners/ownerDetails.html
@@ -1,0 +1,80 @@
+<!DOCTYPE html>
+
+<html xmlns:th="https://www.thymeleaf.org"
+      th:replace="~{fragments/layout :: layout (~{::body},'owners')}">
+
+<body>
+
+
+<h2>Owner Information</h2>
+
+    <!--/*@thymesVar id="owner" type="com.osyrs.sfgpetclinic.model.Owner"*/-->
+
+<table class="table table-striped" th:object="${owner}">
+    <tr>
+        <th>Name</th>
+        <td><b th:text="*{firstName + ' ' + lastName}"></b></td>
+    </tr>
+    <tr>
+        <th>Address</th>
+        <td th:text="*{address}"></td>
+    </tr>
+    <tr>
+        <th>City</th>
+        <td th:text="*{city}"></td>
+    </tr>
+    <tr>
+        <th>Telephone</th>
+        <td th:text="*{telephone}"></td>
+    </tr>
+</table>
+
+<a th:href="@{__${owner.id}__/edit}" class="btn btn-primary">Edit
+    Owner</a>
+<a th:href="@{__${owner.id}__/pets/new}" class="btn btn-primary">Add
+    New Pet</a>
+
+<br />
+<br />
+<br />
+<h2>Pets and Visits</h2>
+
+<table class="table table-striped">
+
+    <tr th:each="pet : ${owner.pets}">
+        <td valign="top">
+            <dl class="dl-horizontal">
+                <dt>Name</dt>
+                <dd th:text="${pet.name}"></dd>
+                <dt>Birth Date</dt>
+                <dd
+                        th:text="${#temporals.format(pet.birthDate, 'yyyy-MM-dd')}"></dd>
+                <dt>Type</dt>
+                <dd th:text="${pet.petType}"></dd>
+            </dl>
+        </td>
+        <td valign="top">
+            <table class="table-condensed">
+                <thead>
+                <tr>
+                    <th>Visit Date</th>
+                    <th>Description</th>
+                </tr>
+                </thead>
+                <tr th:each="visit : ${pet.visits}">
+                    <td th:text="${#temporals.format(visit.date, 'yyyy-MM-dd')}"></td>
+                    <td th:text="${visit?.description}"></td>
+                </tr>
+                <tr>
+                    <td><a th:href="@{__${owner.id}__/pets/__${pet.id}__/edit}">Edit Pet</a></td>
+                    <td><a th:href="@{__${owner.id}__/pets/__${pet.id}__/visits/new}">Add Visit</a></td>
+                </tr>
+            </table>
+        </td>
+    </tr>
+
+</table>
+
+</body>
+
+</html>

--- a/pet-clinic-web/src/main/resources/templates/owners/ownersList.html
+++ b/pet-clinic-web/src/main/resources/templates/owners/ownersList.html
@@ -1,0 +1,61 @@
+<!DOCTYPE html>
+
+<html xmlns:th="https://www.thymeleaf.org" th:replace="~{fragments/layout :: layout (~{::body},'owners')}">
+
+<body>
+
+<h2>Owners</h2>
+
+<table id="owners" class="table table-striped">
+  <thead>
+  <tr>
+    <th style="width: 150px;">Name</th>
+    <th style="width: 200px;">Address</th>
+    <th>City</th>
+    <th style="width: 120px">Telephone</th>
+    <th>Pets</th>
+  </tr>
+  </thead>
+  <tbody>
+  <tr th:each="owner : ${listOwners}">
+    <td>
+      <a th:href="@{/owners/__${owner.id}__}" th:text="${owner.firstName + ' ' + owner.lastName}"/></a>
+    </td>
+    <td th:text="${owner.address}"/>
+    <td th:text="${owner.city}"/>
+    <td th:text="${owner.telephone}"/>
+    <td><span th:text="${#strings.listJoin(owner.pets, ', ')}"/></td>
+  </tr>
+  </tbody>
+</table>
+<div th:if="${totalPages > 1}">
+  <span>Pages:</span>
+  <span>[</span>
+  <span th:each="i: ${#numbers.sequence(1, totalPages)}">
+      <a th:if="${currentPage != i}" th:href="@{'/owners?page=' + ${i}}">[[${i}]]</a>
+      <span th:unless="${currentPage != i}">[[${i}]]</span>
+    </span>
+  <span>]&nbsp;</span>
+  <span>
+      <a th:if="${currentPage > 1}" th:href="@{'/owners?page=1'}" title="First"
+         class="fa fa-fast-backward"></a>
+      <span th:unless="${currentPage > 1}" title="First" class="fa fa-fast-backward"></span>
+    </span>
+  <span>
+      <a th:if="${currentPage > 1}" th:href="@{'/owners?page=__${currentPage - 1}__'}" title="Previous"
+         class="fa fa-step-backward"></a>
+      <span th:unless="${currentPage > 1}" title="Previous" class="fa fa-step-backward"></span>
+    </span>
+  <span>
+      <a th:if="${currentPage < totalPages}" th:href="@{'/owners?page=__${currentPage + 1}__'}" title="Next"
+         class="fa fa-step-forward"></a>
+      <span th:unless="${currentPage < totalPages}" title="Next" class="fa fa-step-forward"></span>
+    </span>
+  <span>
+      <a th:if="${currentPage < totalPages}" th:href="@{'/owners?page=__${totalPages}__'}" title="Last"
+         class="fa fa-fast-forward"></a>
+      <span th:unless="${currentPage < totalPages}" title="Last" class="fa fa-step-forward"></span>
+    </span>
+</div>
+</body>
+</html>

--- a/pet-clinic-web/src/test/java/com/osyrs/sfgpetclinic/controllers/OwnerControllerTest.java
+++ b/pet-clinic-web/src/test/java/com/osyrs/sfgpetclinic/controllers/OwnerControllerTest.java
@@ -13,7 +13,8 @@ import org.springframework.test.web.servlet.setup.MockMvcBuilders;
 
 import java.util.Set;
 
-import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.*;
+import static org.mockito.ArgumentMatchers.anyLong;
 import static org.mockito.Mockito.verifyNoInteractions;
 import static org.mockito.Mockito.when;
 import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
@@ -62,5 +63,13 @@ class OwnerControllerTest {
                 .andExpect(status().isOk())
                 .andExpect(view().name("owners/index"))
                 .andExpect(model().attribute("owners",hasSize(2)));
+    }
+    @Test
+    void displayOwner() throws Exception {
+        when(service.findById(anyLong())).thenReturn(Owner.builder().id(1L).build());
+        mockMvc.perform(get("/owners/123"))
+                .andExpect(status().isOk())
+                .andExpect(view().name("owners/ownerDetails"))
+                .andExpect(model().attribute("owner",hasProperty("id",is(1L))));
     }
 }

